### PR TITLE
Continuously build default-cpu and default-gpu dockerfiles

### LINKED ISSUE #2753: Continuously build default-cpu and default-gpu dockerfiles
It looks like these dockerfiles haven't been updated in a year—it'd be nice if they could be built whenever a change is made.

### PR DESCRIPTION
**Pull request by jzwang43**
### Reasons for making this change

Add a GHA workflow that runs on master whenever the default-cpu/gpu dockerfiles are changed. This should build and push the images to docker hub.

### Related issues

fixes #2753 

### Checklist

* [x] I've added a screenshot of the changes, if this is a frontend change
* [x] I've added and/or updated tests, if this is a backend change
* [x] I've run the [pre-commit.sh](https://github.com/codalab/codalab-worksheets/blob/master/pre-commit.sh) script
* [x] I've updated [docs](https://github.com/codalab/codalab-worksheets/tree/master/docs), if needed

### CODE REVIEW

File: .github/workflows/docker.yml [hunk 1]
```diff
@@ -0,0 +1,35 @@
+name: Push Default
+on:
+  # push:
+  #   branches: [ master ]
+  push:
```
Comment (jzwang43) [inline .github/workflows/docker.yml]: Don't want to accidentally trigger a push so the it's commented out, will change back to the correct command before merging.

File: .github/workflows/docker.yml [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+    branches: [ auto-build ]
+    # paths: ['docker_config/dockerfiles/default-*']
+
+jobs:
+  build_default:
+    name: Build Default Images
```
Comment (epicfaace) [inline .github/workflows/docker.yml]:
```suggestion
    name: Build and push default worker images
```

File: .github/workflows/docker.yml (hunk 1 above)
Comment (epicfaace) [reply .github/workflows/docker.yml]: Not sure what you meant -- this workflow is already being triggered in this PR because of the auto-build condition you added. It's ok to trigger a push because it's being built with the "auto_build" tag rather than the "latest" tag (which is what is used by default).

File: .github/workflows/docker.yml [hunk 3]
```diff
@@ -0,0 +1,35 @@
+name: Push Default
```
Comment (epicfaace) [inline .github/workflows/docker.yml]: Rename this to build_worker_images.yml

Comment (epicfaace) [inline .github/workflows/docker.yml]:
```suggestion
name: Build default worker images
```

File: .github/workflows/docker.yml [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+on:
+  # push:
+  #   branches: [ master ]
+  push:
+    branches: [ auto-build ]
+    # paths: ['docker_config/dockerfiles/default-*']
```
Comment (epicfaace) [inline .github/workflows/docker.yml]: We should also trigger this on a release (see release.yml), in which case it should build and push these images with tag "latest".

**Comment by epicfaace** [comment]
It looks like it only takes 40 minutes to build both images in parallel. Because of that, here's a proposal about when we can run the workflow:
- Run the workflow on PRs, but only whenever the dockerfiles are changed. The image should be pushed with a tag equal to the branch name so that future builds can be cached from that branch.
- Run the workflow on master, but only whenever the dockerfiles are changed. The image should be pushed with a tag equal to "master".
- Run the workflow on release, but only whenever the dockerfiles are changed. The image should be pushed with a tag equal to "latest".

What do you think @jzwang43 ? (also cc'ing @percyliang on this)

**Comment by jzwang43** [comment]
@epicfaace I think we can run on all of the occasions you mentioned since the build time is not too long. Right now I have it run when pushed to master and release. Not sure if running on PRs are needed since if they can build locally then it should have no issues building with git action.

File: .github/workflows/docker.yml (hunk 1 above)
Comment (jzwang43) [reply .github/workflows/docker.yml]: I have updated to trigger this on master and release.

File: .github/workflows/docker.yml (hunk 3 above)
Comment (jzwang43) [reply .github/workflows/docker.yml]: Updated

File: .github/workflows/docker.yml (hunk 4 above)
Comment (jzwang43) [reply .github/workflows/docker.yml]: I updated the code so the workflow is run on PRs, push to masters, and release.

File: .github/workflows/build_worker_images.yml [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+          restore-keys: |
+            pip-
+      - run: pip install -r requirements.txt
+      - run: python3 codalab_service.py build ${SERVICE} $([ -z "${CODALAB_DOCKER_USERNAME}" ] || echo "--push")
+        env:
+          CODALAB_DOCKER_USERNAME: ${{ secrets.CODALAB_DOCKER_USERNAME }}
+          CODALAB_DOCKER_PASSWORD: ${{ secrets.CODALAB_DOCKER_PASSWORD }}
+          # Gives us the branch name of the PR if on a pull_request-triggered build,
+          # otherwise, "master" if on a push-triggered build
+          VERSION: ${{ github.head_ref || 'master' }}
```
Comment (epicfaace) [inline .github/workflows/build_worker_images.yml]: We need to ensure we run an additional build step in which we set VERSION to "latest" when this build is triggered by a release.

File: .github/workflows/build_worker_images.yml [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+          path: ~/.cache/pip
+          key: pip-${{ hashFiles('requirements.txt') }}
+          restore-keys: |
+            pip-
+      - run: pip install -r requirements.txt
+      - run: python3 codalab_service.py build ${SERVICE} $([ -z "${CODALAB_DOCKER_USERNAME}" ] || echo "--push")
+        env:
+          CODALAB_DOCKER_USERNAME: ${{ secrets.CODALAB_DOCKER_USERNAME }}
+          CODALAB_DOCKER_PASSWORD: ${{ secrets.CODALAB_DOCKER_PASSWORD }}
+          # Gives us the branch name of the PR if on a pull_request-triggered build,
```
Comment (epicfaace) [inline .github/workflows/build_worker_images.yml]: Can you update this comment to reflect what VERSION means now (including "latest")?

File: .github/workflows/build_worker_images.yml [hunk 7]
```diff
@@ -0,0 +1,38 @@
+name: Build default worker images
+on:
```
Comment (epicfaace) [inline .github/workflows/build_worker_images.yml:5]: Can you add a comment here explaining when the default worker images are built, and why this decision was made?

File: docker_config/dockerfiles/Dockerfile.default-cpu [hunk 8]
```diff
@@ -64,9 +67,9 @@ ENV SCALA_HOME /usr/share/java
 
 ## Python packages
 
-RUN pip2 install -U pip
+RUN python -m pip install --upgrade "pip < 21.0"
```
Comment (epicfaace) [inline docker_config/dockerfiles/Dockerfile.default-cpu:68]: Why this change?

Comment (jzwang43) [reply docker_config/dockerfiles/Dockerfile.default-cpu:68]: Newer pip dropped the support for python 2, the installation will fail.

File: .github/workflows/build_worker_images.yml (hunk 6 above)
Comment (jzwang43) [reply .github/workflows/build_worker_images.yml]: Comments added, thanks for pointing out.

File: docker_config/dockerfiles/Dockerfile.default-cpu (hunk 8 above)
Comment (epicfaace) [reply docker_config/dockerfiles/Dockerfile.default-cpu:68]: Let's add a comment so it's clear about that

File: docker_config/dockerfiles/Dockerfile.default-cpu [hunk 9]
```diff
@@ -1,7 +1,10 @@
 FROM ubuntu:xenial
 MAINTAINER CodaLab Team "codalab.worksheets@gmail.com"
 
+RUN apt-get update && apt-get install -y \
+    software-properties-common
```
Comment (epicfaace) [inline docker_config/dockerfiles/Dockerfile.default-cpu:5]: Why is this change needed? Also, why did we remove python-software-properties in default-gpu but not in this file?

File: .github/workflows/build_worker_images.yml [hunk 10]
```diff
@@ -0,0 +1,42 @@
+name: Build default worker images
+# Default images are built and pushed when pushed to master, release, and pull request, 
+# and only when the corresponding dockerfiles are changed.
+# This workflow ensures the worker images are up-to-date on docker hub.
```
Comment (epicfaace) [inline .github/workflows/build_worker_images.yml]:
```suggestion
# This workflow ensures the worker images are up-to-date on Docker Hub.
```

File: docker_config/dockerfiles/Dockerfile.default-cpu (hunk 9 above)
Comment (jzwang43) [reply docker_config/dockerfiles/Dockerfile.default-cpu:5]: The step after this `add-apt-repository` will fail if we don't install software-properties-common first.

Comment (jzwang43) [reply docker_config/dockerfiles/Dockerfile.default-cpu:5]: Nevermind I think apt install cannot find python-software-properties. software-properties-common package is an alternate one for python-software-properties and should work fine.

**Review by epicfaace — APPROVED** [review]
